### PR TITLE
Add BeamSearch operator for GPT-2 decoding

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: 

Add BeamSearch operator for GPT-2 decoding. 
- Only CPU implementation is included in this change. CUDA implementation will be added later.
- Only support GPT-2 model right now. It is possible to extend it to support encoder-decoder model in the future.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/beam_earch.cc [hunk 1]
```diff
@@ -0,0 +1,937 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/beam_earch.cc:1]: typo in name of the file beam_search.cc

File: onnxruntime/python/tools/transformers/convert_beam_search.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        helper.make_attribute("pad_token_id", pad_token_id),
+        helper.make_attribute("no_repeat_ngram_size", args.no_repeat_ngram_size),
+        helper.make_attribute("early_stopping", 1 if args.early_stopping else 0),
+        helper.make_attribute("body", model.graph),
+    ])
+
+    from onnx import TensorProto
+
+    # graph inputs
+    input_ids = helper.make_tensor_value_info('input_ids', TensorProto.INT32, ['batch_size', 'sequence_length'])
```
Comment (viboga) [inline onnxruntime/python/tools/transformers/convert_beam_search.py:243]: is this assumption correct? I see gpt2 has int64 when converted to onnx.

File: onnxruntime/contrib_ops/cpu/transformers/beam_earch.cc (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_earch.cc]: good catch. renamed.

File: onnxruntime/python/tools/transformers/convert_beam_search.py (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/convert_beam_search.py:244]: This is input for beam search op, see the operator spec.
gpt2 subgraph indeed use int64 input and output so data cast is needed.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+
+  // Add beam score to next token scores. Corresponding python code is like:
+  //    next_token_scores = next_token_scores + beam_scores[:, None].expand_as(next_token_scores)
+  // TODO: use thread pool to parrellel
+  int offset = 0;
+  int batch_beam_index = 0;
+  for (int i = 0; i < parameters_->batch_size; i++) {
+    for (int j = 0; j < parameters_->num_beams; j++, batch_beam_index++) {
+      for (int k = 0; k < parameters_->vocab_size; k++, offset++) {
+        next_token_scores[offset] += beam_state.beam_scores[batch_beam_index];
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:409]: Here, I understand we are updating the scores of next tokens from previous beams. Since this is a summation for all the logits - topK should not be affected by this? Doing this after topK is more efficient?

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+  TensorShape next_token_scores_shape(&next_token_scores_dims[0], 2);
+  auto element_type = DataTypeImpl::GetType<T>();
+  OrtValue next_token_scores_value;
+  Tensor::InitOrtValue(element_type, next_token_scores_shape, next_token_scores.data(), allocator->Info(), next_token_scores_value);
+  const Tensor& input = next_token_scores_value.Get<Tensor>();
+
+  const int axis = 1;
+  const unsigned top_k = static_cast<unsigned>(2 * parameters_->num_beams);
+  const bool largest = true;
+  const bool sorted = true;  // results returned in sorted order.
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:437]: Why does it need to be sorted?

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    processor_list_.push_back(repetition_penalty_processor_.get());
+  }
+
+  if (parameters.no_repeat_ngram_size > 0) {
+    no_repeat_ngram_processor_ = std::make_unique<NoRepeatNGramLogitsProcessor<T>>(parameters.no_repeat_ngram_size);
+    processor_list_.push_back(no_repeat_ngram_processor_.get());
+  }
+
+  if (!parameters.vocab_mask.empty()) {
+    vocab_mask_processor_ = std::make_unique<VocabMaskLogitsProcessor<T>>(parameters.vocab_mask);
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:161]: vocab mask supplied is used to zero out unrelated vocabs.
Semantically, shouldn't this precede any other processing - repetition or no_repeat?

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+void VocabMaskLogitsProcessor<T>::Process(const ISequences* /*sequences*/,
+                                          NextTokenScores<T>& next_token_scores) {
+  assert(!vocab_mask_.empty());
+
+  // Process vocabulary mask and set tokens with mask value 0 to -inf.
+  T* p = next_token_scores.scores.data();
+  // next_token_scores shape (batch_size * num_beams, vocab_size)
+  // vocab_mask shape (vocab_size). TODO: support shape (batch_size, vocab_size)
+  for (int i = 0; i < next_token_scores.batch_beam_size; i++) {
+    for (int j = 0; j < next_token_scores.vocab_size; j++, p++) {
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:134]: We are applying same vocab mask for all the batches.
This means all the sequences have same vocab mask. One use case I think of vocab mask is where it is prepared by tokenzier to mark the relevant tokens in next inference(the first one). Each batch can have different seq_len and different vocab mask.
Is this done to address a diff case?

File: onnxruntime/contrib_ops/cpu/transformers/sequences.cc [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+namespace onnxruntime {
+namespace contrib {
+namespace transformers {
+
+void Sequences::Init(AllocatorPtr allocator, const OrtValue& input_ids, int batch_beam_size, int sequence_length, int max_length) {
+  size_t sequences_size = SafeInt<size_t>(batch_beam_size) * max_length;
+  size_t buffer_size = sequences_size + sequences_size;
+  gsl::span<int64_t> buffer = AllocateBuffer<int64_t>(allocator, sequences_space_buffer_, buffer_size, true, static_cast<int64_t>(0));
+
+  sequences[0] = buffer.subspan(0, sequences_size);
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/sequences.cc:12]: Are 2 buffers needed? with max_length known or with default max_length - could we append to the same buffer based on last position?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 8]
```diff
@@ -568,6 +568,139 @@ void DecoderAttentionTypeAndShapeInference(ONNX_NAMESPACE::InferenceContext& ctx
...
+      scores_shape.add_dim()->set_dim_value(batch_size);
+      scores_shape.add_dim()->set_dim_value(num_beams_value);
+      scores_shape.add_dim();  // vocab_size is unknown
+      updateOutputShape(ctx, 2, scores_shape);
+    }
+  }
+}
+
+void RegisterTextGenerationSchemas() {
+  ONNX_CONTRIB_OPERATOR_SCHEMA(BeamSearch)
```
Comment (viboga) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:664]: Could you add README for beam search detailing out implementation? I see this is a big change and due to the number of operators involved and two levels of graph, this would help in tracking and extending the work easily.

Comment (viboga) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:664]: For instance, we can add for what vocab mask should (shouldn't) be used.

File: onnxruntime/test/python/transformers/test_beam_search.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import pytest
+
+from parity_utilities import find_transformers_source
+if find_transformers_source():
+    from convert_beam_search import main as run
+else:
+    from onnxruntime.transformers.convert_beam_search import main as run
+
+
+class TestBeamSearch(unittest.TestCase):
```
Comment (viboga) [inline onnxruntime/test/python/transformers/test_beam_search.py:20]: I believe we need test cases at operator level too, just like others.

**Comment by viboga** [comment]
Looks like its much behind master.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+
+  CHECK_SCALAR_INPUT(num_beams, 3, true);
+
+  CHECK_SCALAR_INPUT(num_return_sequences, 4, true);
+
+  CHECK_SCALAR_INPUT(temperature, 5, true);
+
+  CHECK_SCALAR_INPUT(length_penalty, 6, true);
+
+  ORT_RETURN_IF(parameters_->num_return_sequences > parameters_->num_beams, "'num_return_sequences' has to be smaller or equal to 'num_beams'.");
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:323]: trying to recall justification over this: every new iteration in beam search doesn't have to be extracting num_return_sequences. The search can go on with num_beams among the non-terminated sequences.

This can be done as an extention/improvement too.

File: onnxruntime/python/tools/transformers/convert_beam_search.py [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+    print("Test ONNX model and bream search with onnxruntime...")
+
+    ort_session = create_ort_session(args.output, args.use_gpu)
+
+    vocab_mask = np.ones((vocab_size), dtype=np.int32)
+    if use_vocab_mask:
+        for bad_word_id in bad_words_ids:
+            vocab_mask[bad_word_id] = 0
+
+    inputs = {
```
Comment (viboga) [inline onnxruntime/python/tools/transformers/convert_beam_search.py:378]: the arguments passed to conversion script are used here to test. However, the graph itself would only have defaults since these are inputs. We can remove these optional values to parameters

[--temperature TEMPERATURE]
[--length_penalty LENGTH_PENALTY] 
[--repetition_penalty REPETITION_PENALTY]

unless I am missing something.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:409]: Topk is on each batch. The added scores is per beam so this shall be done before top-k otherwise the result will be different.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 4 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:437]: we need to process best results first (in some case, we can stop early so top_k shall be sorted).

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 5 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:161]: Here, we uses the same order as in huggingface transformer to get parity result:

https://github.com/huggingface/transformers/blob/e513c16e82221e0a150b97a31f2c3c864913c454/src/transformers/generation_utils.py#L610-L639

NoBadWordsLogitsProcessor is after RepetitionPenaltyLogitsProcessor and NoRepeatNGramLogitsProcessor.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 6 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:134]: That's the reason I added a comment above the code "TODO: support shape (batch_size, vocab_size)".

There are two scenarios: one case is to used as bad word list as in huggingface transformers, that list is global. Another case is per batch mask. That can be done in another PR.

File: onnxruntime/contrib_ops/cpu/transformers/sequences.cc (hunk 7 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/sequences.cc:12]: For example, currently there are two sequences: "a b",  "c d". When we append next tokens (for example, "e" and "f" are top-2), the best sequences might be "a b e", "a b f" after top-k sorting.

That means we need to keep track of sequences of previous steps ("a b", "c d"), and also current steps ("a b e", "a b f"). One way to do that is to have two buffers, one for previous step, and another for current step.

If we do not use two buffers, we have to record the beam index picked in each step, then reverse trace to find the whole sequences. That will be more complicated.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 8 above)
Comment (tianleiwu) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:664]: There is some description of each input and each attribute in operator definition. Is that good enough?

File: onnxruntime/test/python/transformers/test_beam_search.py (hunk 9 above)
Comment (tianleiwu) [reply onnxruntime/test/python/transformers/test_beam_search.py:20]: This is operator level test (by comparing the parity with huggingface transformer).

We can probably also add a small model and hard code the expected result, and enable test in CI pipeline later. There's a comment "#TODO: use a smaller model and enable tests in CI pipeline" in this file.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 10 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:323]: For each batch, we only keep track of num_beams results, so the number of results (per batch) to return cannot exceed num_beams.

File: onnxruntime/python/tools/transformers/convert_beam_search.py (hunk 11 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/convert_beam_search.py:378]: To avoid confusion, how about using initializers to store these parameters in ONNX graph?

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 6 above)
Comment (viboga) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:134]: Internal TODOs are good for devs. My question is more on the restriction this brings in. Are you documenting this? What can we provide as reference to users?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 8 above)
Comment (viboga) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:664]: Lets discuss on this a bit more offline, we need to have more documentation of our features and coverage.

File: onnxruntime/test/python/transformers/test_beam_search.py (hunk 9 above)
Comment (viboga) [reply onnxruntime/test/python/transformers/test_beam_search.py:20]: I meant .cc

File: onnxruntime/python/tools/transformers/convert_beam_search.py (hunk 11 above)
Comment (viboga) [reply onnxruntime/python/tools/transformers/convert_beam_search.py:378]: why does convert_beam_search.py need these parameters passed in?

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 10 above)
Comment (viboga) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:323]: Beam search is only to expand the search space. Say, there is no beam search, the very first iteration can still return num_sequences with highest scored logits.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 4 above)
Comment (viboga) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:437]: I see, why 2*num_beams though?

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:437]: some processor (like vocab mask) will filter some "bad words". That means we shall get top_k > num_beams. 2 * num_beams seems to be a reasonable value chosen in huggingface code that we followed.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 6 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:134]: Most TODOs shall be resolved before next release 1.11. For reference, the definition of the operator has shape information for each input.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 10 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:323]: Each iteration only generate one word, and there is no limit on token level.

The limitation is on number of sequences. I have not seen a case with num_return_sequences > num_beams.

File: onnxruntime/python/tools/transformers/convert_beam_search.py (hunk 11 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/convert_beam_search.py:378]: This script generates ONNX model, and it also serves an example of how to use ONNX model.

So we exposed the parameters so that user could try different values to see its impact. Currently, those parameters are used in inference only, and not store in ONNX model. In the future, we could add an option that user to choose which parameters needed to be store as initializers in ONNX model so that they need not specify in inputs for inference.

**Review by viboga — APPROVED** [review]

**Review by gwang-msft — APPROVED** [review]
Sign off for kernel hash change
